src/tsjson.test.ts

```diff
@@ -308,6 +308,175 @@ describe("More involved tests with objects", () => {
   });
 });
 
+describe("additionalProperties", () => {
```

Comment:
@ostrowr these tests are better at highlighting what I'm trying to fix. The type checking is a little off when mixing `properties` with `additionalProperties`. These tests that I have added fail both with and without my changes to `json-schema.ts`. I.e. the issue exists in the current `master` branch, and my changes to `json-schema.ts` do not fix them.

It seems very difficult to get better type checking for this scenario, so I'd appreciate any pointers or suggestions you may have and I'll be happy to continue working on this.

Thanks!

Reply:
Unrelated, but I wanted to say anyway: I found the `expectType()` function quite clever! That was a nifty way to quickly and easily do typing tests

Reply:
Thanks!

I'm going to take a closer look at this PR tonight/tomorrow – appreciate the new tests.

Reply:
You know this code way better than I do, but my guess is that the fix will require combining `PropertiesConstraint` and `AdditionalPropertiesConstraint` the same was that `AdditionalItems` is included in `ItemsConstraint`

Ex.
```
PropertiesConstraint<
  Properties extends { [k: string]: SchemaLike } | undefined,
  Required extends readonly (keyof Properties)[] | undefined,
  AdditionalProperties extends SchemaLike | undefined // add this, and remove AdditionalPropertiesConstraint
> = // stuff
```

I don't know if it's possible in Typescript to define an `object like this:
```
type PropertiesConstraint  = {
    [P in keyof Properties]: Properties[P][typeof InternalTypeSymbol],
    [k: /* all other properties */]: unknown
}
```
so for `additionalProperties: S(true)` the best might be that it gives you `{ [k: string]: unknown }` and loses all type information for `properties`. But for `additionalProperties: S(false)`, it's probably possible to have it revert back to the `properties` schemas.

Reply:
Yeah, it seems like a perfect typing is blocked by https://github.com/microsoft/TypeScript/issues/17867

Individual property access seems to be correctly typed, it's just the `expectType<Validated<typeof schema>>(...)` checks that fail.

This may just be impossible at this point, unfortunately.